File: js/harness.ts
/** Batch harness for conformance/run.py. Protocol: stdin first line = N, then N
 * items; stdout = N results. encode: hex -> ba64 text. decode: ba64 -> hex or
 * "!CODE". See python/harness.py for the full protocol description. */
import { Buffer } from "node:buffer";
import { encode, decode, Ba64Error } from "./ba64.ts";

const mode = process.argv[2];
const chunks: Buffer[] = [];
process.stdin.on("data", (c) => chunks.push(c));
process.stdin.on("end", () => {
  const data = Buffer.concat(chunks).toString("utf8").split("\n");
  const n = parseInt(data[0], 10);
  const items = data.slice(1, 1 + n);
  const out: string[] = [];
  for (const item of items) {
    if (mode === "encode") {
      out.push(encode(new Uint8Array(Buffer.from(item, "hex"))));
    } else {
      try {
        out.push(Buffer.from(decode(item)).toString("hex"));
      } catch (e) {
        if (e instanceof Ba64Error) out.push("!" + e.code);
        else throw e;
      }
    }
  }
  process.stdout.write(out.join("\n"));
});

File: python/harness.py
#!/usr/bin/env python3
"""Batch harness for cross-language conformance (conformance/run.py).

Protocol (stdin/stdout), line-based, count-prefixed so empty items are
unambiguous:

  stdin  : first line = N, then N item lines
  stdout : N result lines

  encode : each item is hex(input bytes)     -> result is the ba64 text
  decode : each item is a ba64 text          -> result is hex(bytes) or "!CODE"
"""

import sys

from ba64 import decode, encode, Ba64Error


def main():
    mode = sys.argv[1]
    data = sys.stdin.buffer.read().decode("utf-8", "surrogateescape").split("\n")
    n = int(data[0])
    items = data[1:1 + n]
    out = []
    for item in items:
        if mode == "encode":
            out.append(encode(bytes.fromhex(item)))
        else:
            try:
                out.append(decode(item).hex())
            except Ba64Error as e:
                out.append("!" + e.code)
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()
